### DISCUSSION TITLE
distinct_on() multiple columns support

### DISCUSSION THREAD

**Discussion by Ticki84** [diesel-rs/diesel#3212] [Ideas]
Let's take a pretty basic SQL example to query latest objects from table `tab` by date with a distinct on two columns:

```
SELECT DISTINCT ON (field1, field2) *
    FROM tab
    ORDER BY field1, field2, date DESC;
```

Which is easily translated into this code using the Diesel ORM:

```
use super::schema::tab;

tab::table
    .distinct_on((tab::field1, tab::field2))
    .order((tab::field1, tab::field2, tab::date.desc()));
```

But yet, only single value (i.e. one column) are supported as `distinct_on()` param. It would be great to support multiple columns in  `distinct_on()`.

In this simple example, it isn't a big deal because you could simply write the SQL query. But for big queries, you could use these to filter out results in a same request. Another workaround is to create a view of the first SQL select, then create a table for this view using [this trick](https://deterministic.space/diesel-view-table-trick.html), but the amount of work required is high for such a simple request.

**Comment by weiznich** [comment]
This is essentially a duplicate of https://github.com/diesel-rs/diesel/issues/3020

That written: I would accept a PR for this, as long as it includes tests + compile tests. For me personally this is not a priority issue, so I likely won't work on this in the next time.

**Reply by lochetti** [reply]
I would like to try to help here with this evolution.
That said, I have never contributed to the diesel project source code and I really don't know if this problem would be a good first problem to solve (and if it is quite difficult or not).
Do you have any opinions or suggestions?

**Reply by weiznich** [reply]
I did not investigate in depth yet what's required to enable the presented usecase. That remains an open question and that's likely the main work that is required here. If someone is willing to do that work that would be great. Otherwise this is a pretty well contained thing in diesels code. To solve it you likely require a good understanding of how traits work in rust and how to write generic impls.

From a short look at the code I can say that these (https://github.com/diesel-rs/diesel/blob/fa33e38e53127ad5e42d5b5566028e29b96fd605/diesel/src/pg/query_builder/distinct_on.rs#L18-L52) impls are not general enough to cover the case presented above. That can be likely fixed by adding additional impls there or by generalizing existing impls. I would likely want to see additional test cases for the newly added cases here (https://github.com/diesel-rs/diesel/blob/fa33e38e53127ad5e42d5b5566028e29b96fd605/diesel_tests/tests/distinct.rs#L1-L100) and additional compile fail tests here (https://github.com/diesel-rs/diesel/blob/fa33e38e53127ad5e42d5b5566028e29b96fd605/diesel_compile_tests/tests/fail/distinct_on_requires_matching_order_clause.rs#L1-L90) for rejected cases.
